src/components/Approval/Approval.tsx

```diff
@@ -0,0 +1,173 @@
+import { Button, Divider, Flex, Image, Link, SkeletonCircle } from '@chakra-ui/react'
+import { ChainTypes, SwapperType } from '@shapeshiftoss/types'
+import { useEffect, useRef, useState } from 'react'
+import { CountdownCircleTimer } from 'react-countdown-circle-timer'
+import { useFormContext } from 'react-hook-form'
+import { useHistory, useLocation } from 'react-router-dom'
+import { MiddleEllipsis } from 'components/MiddleEllipsis/MiddleEllipsis'
+import { Row } from 'components/Row/Row'
+import { SlideTransition } from 'components/SlideTransition'
+import { RawText, Text } from 'components/Text'
+import { useSwapper } from 'components/Trade/hooks/useSwapper/useSwapper'
+import { TradeState } from 'components/Trade/Trade'
+import { useWallet } from 'context/WalletProvider/WalletProvider'
+import { useLocaleFormatter } from 'hooks/useLocaleFormatter/useLocaleFormatter'
+import { bn } from 'lib/bignumber/bignumber'
+import { theme } from 'theme/theme'
+
+type ApprovalParams = {
+  fiatRate: string
+}
+
+export const Approval = () => {
+  const history = useHistory()
+  const location = useLocation<ApprovalParams>()
+  const approvalInterval: { current: NodeJS.Timeout | undefined } = useRef()
+  const [approvalTxId, setApprovalTxId] = useState<string>()
+  const { fiatRate } = location.state
+
+  const {
+    getValues,
+    handleSubmit,
+    formState: { errors, isSubmitting }
+  } = useFormContext<TradeState<ChainTypes, SwapperType>>()
+  const { approveInfinite, checkApprovalNeeded, buildQuoteTx } = useSwapper()
+  const {
+    number: { toCrypto, toFiat }
+  } = useLocaleFormatter({ fiatType: 'USD' })
+  const {
+    state: { wallet }
+  } = useWallet()
+  const { quote, sellAsset, fees } = getValues()
+  const fee = fees?.chainSpecific?.approvalFee
+  const symbol = sellAsset.currency?.symbol
+
+  const approve = async () => {
+    if (!wallet) return
+    const txId = await approveInfinite(wallet)
+    if (!txId) return
+    setApprovalTxId(txId)
+    const interval = setInterval(async () => {
+      const approvalNeeded = await checkApprovalNeeded(wallet)
+      if (approvalNeeded) return
+      approvalInterval.current && clearInterval(approvalInterval.current)
+      if (!sellAsset.amount) return
+      if (!quote) return
+      const result = await buildQuoteTx({
+        wallet,
+        sellAsset: quote?.sellAsset,
+        buyAsset: quote?.buyAsset,
+        amount: sellAsset?.amount
+      })
+
+      if (result?.success) {
+        history.push({ pathname: '/trade/confirm', state: { fiatRate } })
+      } else {
+        history.push('/trade/input')
+      }
+    }, 5000)
+    approvalInterval.current = interval
+  }
```

Comment:
No try/catch can result in an `UnhandlePromiseRejection` error

Reply:
Someone else's PR includes a "poll" utility function that would simplify this code. You may want to go back a refactor after that gets merged.

Reply:
This is another ticket https://github.com/shapeshift/lib/issues/152